walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,106 @@
 
 package com.google.sps;
 
+import java.util.Comparator;
 import java.util.Collection;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.Arrays;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
-}
+
+    private final ArrayList < TimeRange > options = new ArrayList < TimeRange > ();
+
+    public Collection < TimeRange > query(Collection < Event > events, MeetingRequest request) {
+
+        Collection attendees = request.getAttendees();
+        Collection optionalAttendees = request.getOptionalAttendees();
+
+        // Check if duration is longer than the day. if so, return empty collection.
+        long duration = request.getDuration();
+        if (duration > TimeRange.WHOLE_DAY.duration()) {
+            return Arrays.asList();
+        }
+
+        // If no attendees or optional attendees return collection of one item being the full day
+        if (attendees.isEmpty() && optionalAttendees.isEmpty()) {
+            return Arrays.asList(TimeRange.WHOLE_DAY);
+
+        // If there are optional attendees but no attendees, optional attendees will be treated as attendees
+        } else if (attendees.isEmpty()) {
+            attendees = optionalAttendees;
+        }
+
+        // If no events return the whole day
+        if (events.isEmpty()) {
+            return Arrays.asList(TimeRange.WHOLE_DAY);
+        }
+
+        // Sorts array of event objects by start time in ascending order
+        ArrayList < Event > eventsList = new ArrayList < Event > (events);
+        Collections.sort(eventsList, Event.ORDER_BY_START_TIME);
+        Iterator < Event > eventsListIterator = eventsList.iterator();
+
+        // Initializing variables used in eventsList iterator loop
+        int start = TimeRange.START_OF_DAY;
+        Event conflict = null;
+        Event prevConflict = null;
+
+        while (eventsListIterator.hasNext()) {
+            // If previous conflict end is after current conflict end, replace current conflict with previous conflict
+            conflict = eventsListIterator.next();
+            if (prevConflict != null) {
+                if (prevConflict.getWhen().end() > conflict.getWhen().end()) {
+                    conflict = prevConflict;
```

Comment:
this feels inefficient - you've already evaluated the previous conflict so why is there any reason to reevaluate it again? the naming is also slightly confusing, in that you are calling an event a conflict, when it might not actually conflict? can the code be simplified to make this problem go away?

Reply:
I renamed conflict to event. What exactly is inefficient? I don't get what you mean by evaluate previous conflict